homeassistant/components/buienradar/camera.py

```diff
@@ -0,0 +1,178 @@
+"""Provide animated GIF loops of Buienradar imagery."""
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from typing import Optional
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.components.camera import PLATFORM_SCHEMA, Camera
+from homeassistant.const import CONF_ID, CONF_NAME
+
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from homeassistant.util import dt as dt_util
+
+
+CONF_DIMENSION = 'dimension'
+CONF_DELTA = 'delta'
+
+RADAR_MAP_URL_TEMPLATE = ('https://api.buienradar.nl/image/1.0/'
+                          'RadarMapNL?w={w}&h={h}')
+
+_LOG = logging.getLogger(__name__)
+
+# Maximum range according to docs
+DIM_RANGE = vol.All(vol.Coerce(int), vol.Range(min=120, max=700))
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend({
+        vol.Optional(CONF_DIMENSION): DIM_RANGE,
+        vol.Optional(CONF_DELTA): vol.All(vol.Coerce(float),
+                                          vol.Range(min=0)),
+        vol.Optional(CONF_NAME): cv.string,
+    }))
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up buienradar radar-loop camera component."""
+    c_id = config.get(CONF_ID)
+    dimension = config.get(CONF_DIMENSION) or 512
+    delta = config.get(CONF_DELTA) or 600.0
+    name = config.get(CONF_NAME) or "Buienradar loop"
+
+    async_add_entities([BuienradarCam(hass, name, c_id, dimension, delta)])
+
+
+class BuienradarCam(Camera):
+    """
+    A camera component producing animated buienradar radar-imagery GIFs.
+
+    Rain radar imagery camera based on image URL taken from [0].
+
+    [0]: https://www.buienradar.nl/overbuienradar/gratis-weerdata
+    """
+
+    _dimension = 0          # type: int
+    """ Deadline for image refresh """
+    _deadline = None        # type: Optional[datetime]
+    _name = ""              # type: str
+    _delta = 0.0         # type: float
+    _condition = None       # type: Optional[asyncio.Condition]
+    """ Loading status """
+    _loading = False        # type: bool
+
+    _last_image = None      # type: Optional[bytes]
+    """ last modified HTTP response header"""
+    _last_modified = None   # type: Optional[str]
+
+    def __init__(self, hass, name: str, c_id: Optional[str], dimension: int,
+                 delta: float):
+        """Initialize the component."""
+        super().__init__()
+        self._hass = hass
+        self._name = name
+
+        self._dimension = dimension
+        self._delta = delta
+
+        self._condition = asyncio.Condition(loop=hass.loop)
```

Comment:
If we create the condition before creating the entity, and pass the condition to the entity, we don't need to pass in `hass` to the entity.

We don't need to set `loop` explicitly on the condition if we're in a coroutine.

Reply:
Creating a condition before creating the entity obscures the invariant that the condition is private to this entity. I would like keep the condition object internal because of this (sharing it would cause `notify_all` to notify where it shouldn't) and will document that.

I agree that it is not needed to store a private reference to `hass`, will add an optional parameter for the event loop since it is considered good practice to do so.

You imply an invariant here which was not clear to me from the documentation and other code. Note that this invariant is logical but does not necessarily implicitly hold.

I think that this holds: "async setup methods are called with the event loop of the thread (i.e. `asyncio.get_event_loop()` being equal to hass.loop". If this holds it would really help if part of the platform lifecycle  is documented :)

Reply:
I'm not 💯 on what you mean with invariant. Sorry, I don't have a computer science background.

> You imply an invariant here which was not clear to me from the documentation and other code. Note that this invariant is logical but does not necessarily implicitly hold.

What are you referring to here? `hass` being set on the entity on addition to home assistant or current event loop being used for `loop` by default inside coroutines (inside the event loop)?

Reply:
We can create the condition inside init without passing in hass if we don't set loop explicitly. That's my suggestion after your explanation about condition being private, which makes sense.

Reply:
My explanation was not 100% precise either, I removed an extraneous "implicitly". With invariant I mean "a property that always holds"/"it always is true that ..."

> What are you referring to here? hass being set on the entity on addition to home assistant or
> current event loop being used for loop by default inside coroutines (inside the event loop)?

To me it was not clear that `hass.loop` (in `async_setup_component`) is the same as the default event loop used by coroutines (inside the event loop). It feels logical but would cause potentially cause bugs if it is not. Being unsure I would prefer to pass it in explicitly

Reply:
The default loop in coroutines is the current loop used in the event loop that runs the coroutines. This is part of asyncio library. Since we're in a coroutine in `async_setup_platform` we don't need to care about what loop to use. We just ask asyncio to use the current loop by not setting it explicitly. This is safe.

Reply:
> The default loop in coroutines is the current loop used in the event loop that runs the coroutines. This is part of asyncio library. Since we're in a coroutine in `async_setup_platform` we don't need to care about what loop to use. We just ask asyncio to use the current loop by not setting it explicitly. This is safe.

There is a tiny, tiny caveat there. From my understanding, `asyncio.get_running_loop` gets the current loop. By default asyncio Lock's/Condiotion's/... [use](https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L162) `asyncio.get_event_loop()`.

This _can be_ a different one, and will cause "surprising" issues in that situation (RuntimeError's or difficult to debug situations when there are multiple event loops)... Glad that is not the case here.

Reply:
The [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop) on `asyncio.get_event_loop` doesn't say it can return a different loop than the current one, if I'm reading correctly. If there's no current loop set, it will create a new loop and set it as current. Yeah, there might be cases where this can be a problem. I don't see cases at the moment that I'm afraid should impact us.

Reply:
There's this note in the docs though, just below this [section](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.new_event_loop):
> Note that the behaviour of get_event_loop(), set_event_loop(), and new_event_loop() functions can be altered by setting a custom event loop policy.

Reply:
> Yeah, there might be cases where this can be a problem. I don't see cases at the moment that I'm afraid should impact us.

Exactly! It is one of those 'should not happens' that cause people to be careful. I did a quick grep for this yesterday and the only place in this project (note that libraries can break this...) that could cause issues that I found is usage of `homeassistant.util.async.asyncio_run` which would unset the event loop afterwards.

But in coroutines it still seems to work in practice. See https://gist.github.com/ties/21d75aa31c399e6c9a100232d040f214 for a bunch of test cases :)

> There's this note in the docs though, just below this [section](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.new_event_loop):
> 
> > Note that the behaviour of get_event_loop(), set_event_loop(), and new_event_loop() functions can be altered by setting a custom event loop policy.

The general gist of this is what caused me to be defensive: As a new contributor you see that event loops are commonly passed in explicitly in the project and there are few guarantees.

I'll keep a note that it may be good to add the fact that the event loop returned by `get_event_loop` is assumed to be the correct one to the documentation. Without a custom event loop this is actually what seems to happen.

Reply:
They fixed it in Python 3.5.3 so that get_event_loop called from inside an async function will always return the event loop that is running the async task: https://bugs.python.org/issue28613

(see also https://docs.python.org/3.5/whatsnew/changelog.html#python-3-5-3 and search for get_event_loop)

Reply:
If you see code passing in the event loop, it's because it was most likely written prior to our minimum requirement being raised to 3.5.3

Reply:
Really nice piece of clarification. Thanks! 👍